### ISSUE TITLE
Migrating from J2V8 to Javet/Javenode?

### ISSUE THREAD

**Issue by Awesome1221** [Block-JS/blockjs#2]
Since J2V8 isn't maintained that much lately, and there are better alternatives to work with, I think the actual library should've been deprecated from the project baseline and replaced with Javet/Javenode.

See here: https://www.caoccao.com/Javet/tutorial/migration_guides/migrate_from_j2v8.html

**Comment by TheGreatViolet** [comment]
It's actually funny you should mention this, I've been working on trying to migrate this project to Javet for a little bit now (see branch at https://github.com/Block-JS/blockjs/tree/javet). The migration is done for the most part, I've just hit a couple snags (mainly not being able to import modules without Javet throwing an exception)

**Comment by Awesome1221** [comment]
Yeah, I've noticed some issues with Javet as well, although I don't know if these are the same ones we're talking about. I just made a pull request, you can check it. As of now there's no way to directly pass a File/Path reference to the Javet runtime, as it messes up with the file reading/writing, possibly a bug?

**Comment by TheGreatViolet** [comment]
Well the pull request seems to have fixed one of the issues, that being that whereas before the PR when I'd run the test command `/bjs test` again after having ran it before it wouldn't run again.

However another issue that I had before still remains. When I'd try to import another file/an NPM module using the import/export syntax Javet claims to support, I get an error like this:

```
[10:25:51 INFO]: litteraly1984 issued server command: /bjs test
[10:25:51 WARN]: [com.caoccao.javet.interop.NodeRuntime] 7 V8 callback context object(s) not recycled.
[10:25:51 WARN]: [com.caoccao.javet.interop.NodeRuntime] 7 V8 callback context object(s) not recycled.
[10:25:51 ERROR]: [BlockJS] There was an error with the script: test2.js
[10:25:51 ERROR]: [BlockJS] Error: Cannot find package './test'
[10:25:51 WARN]: [com.caoccao.javet.interop.NodeRuntime] 7 V8 callback context object(s) not recycled.
```

I'm not entirely sure if this is a problem with Javet or if it's perhaps something I'm doing wrong, but I haven't gotten things to work no matter what I try. If anyone wants to take a look I've attached the full logs along with the files I used to test.

#### **`test.js`**
```js
export function test() {
  return 'test';
}
```
#### **`test2.js`**
```js
import { test } from "./test";

console.log(test());
```

[latest.log](https://github.com/Block-JS/blockjs/files/9134272/latest.log)

**Comment by Awesome1221** [comment]
Perhaps it might be related to https://github.com/caoccao/Javet/discussions/4

**Comment by Awesome1221** [comment]
I checked it, and it looks like even after trying to set up a custom module resolver, it doesn't seem to help. When running the command for the first time it seems to be working, however when the V8 context is reused those imported modules appear to still be there despite the fact that V8Runtime.resetContext() is being used. 

>\> bjs test
[INFO]: test
[INFO]: [BlockJS] [STDOUT] function test() {
    return 'test';
}
\> bjs test
[INFO]: test
[ERROR]: [BlockJS] There was an error with the script: test2.js
[ERROR]: [BlockJS] Error: Cannot read the array length because "<local6>" is null
...
[ERROR]: Could not save data net.minecraft.world.entity.raid.PersistentRaid
...
[ERROR]: Failed to save level ./world
...

Seems like it's the same thing https://github.com/Block-JS/blockjs/issues/2#issuecomment-1186707765

**Comment by TheGreatViolet** [comment]
Alright, I was able to get a custom module resolver working in the test suite for the JSExecutionInterface, and it seems to work quite well. All that's left now is to get that resolver working in the JSRunner class

**Comment by TheGreatViolet** [comment]
I ended up with something similar to this. I'm not entirely sure what I did different, but it seems to work when reusing the context:

```java
NodeRuntime createRuntime(File file) throws JavetException {
        JavetEngineConfig config = new JavetEngineConfig();
        config.setJSRuntimeType(JSRuntimeType.Node);

        JavetEnginePool<NodeRuntime> pool = new JavetEnginePool<NodeRuntime>(config);

        NodeRuntime runtime = pool.getEngine().getV8Runtime();

        runtime.setV8ModuleResolver(((v8Runtime, resourceName, iv8Module) -> {
            // Matches the file extension
            Pattern pattern = Pattern.compile("((?<=\\w)\\.[^.]+)$");
            Matcher matcher = pattern.matcher(resourceName);

            String ext = matcher.find() ? matcher.group(1) : "";

            String trueResourceName;

            if (ext.equals("")) {
                trueResourceName = resourceName + ".js";
            } else {
                trueResourceName = resourceName;
            }

            File resourceFile = new File(file.getParentFile(), trueResourceName);

            if (!resourceFile.exists()) {
                // TODO: Resolve module from node_modules
                return null;
            }

            return runtime.getExecutor(resourceFile)
                    .setResourceName(resourceName)
                    .compileV8Module();
        }));

        return runtime;
    }
```

Pretty much all that needs to be done now is resolving any packages from `node_modules`.
